src/producer/transactionManager.js

```diff
@@ -93,21 +209,111 @@ module.exports = ({ logger, cluster, transactionTimeout = 60000 }) => {
     },
 
     /**
-     * Get the current producer id
-     * @returns {number}
+     * Begin a transaction
      */
-    getProducerId() {
-      return producerId
+    beginTransaction() {
+      transactionalGuard()
```

Comment:
I agree that we should aim for good ergonomics, but I'm afraid a transaction object will create confusion around who is responsible for the transaction. At the end of the day it's still the producer that sends messages and advances the state of the transaction. Will be good to have more feedback on the best design from @Nevon et al

Reply:
Given that we get so many questions that are just misunderstandings of the APIs or calling things out of order, I would strongly advocate that if we can make it impossible to misuse the external API, we should make it so, rather than giving the users' a footgun and then yelling at them when they shoot themselves. 😅

> At the end of the day it's still the producer that sends messages and advances the state of the transaction

It sounds to me like the problem then is that the publicly exposed producer interface is the same one that we use internally. The idea behind returning the transaction object was that then there wouldn't be a way to try to send from a transactional producer without having started a transaction, because that API wouldn't be exposed.

Some fantasy example code (ignore the option names, parameters, etc.):

```javascript
const producer = kafka.producer({ transactional: true })
const transaction = await producer.transaction() // this is basically like `beginTransaction`
await transaction.send([ ... ])
await transaction.commit()

producer.send() // `send` is not a function, so there's nothing to call
```

You know more of the details of how the transactional producer works, so maybe this design isn't feasible, but I would really like to avoid us doing the same thing that the Java client does, as I think it's neither user-friendly nor nice for the code, as you end up with all these guards everywhere.

Reply:
Although thinking about it some more, if each producer can only have one transaction in progress at any given time, the user could still try to create more transactions before committing or rolling back the previous one, so I guess it would still be prone to misuse. 🤔